### DISCUSSION TITLE
Replace TOTP with Webauthn/Passkeys?

### DISCUSSION THREAD

**Discussion by eapl-gemugami** [sorenpeter/timeline#45] [Ideas]
A few years ago I implemented TOTP (A 10 digit password changing every 30 seconds) in my PHP projects, such as twtxt-php, as a way to avoid using static passwords (which are problematic in terms of security)

Although I use it, it requires a separate client to generate dynamic passwords (CLI or Mobile app), and honestly I don't see it'll be used broadly by many users.

As Passkeys are becoming more popular, perhaps we could replace TOTP with this more secure standard.
My main concert will be on using an external library like https://github.com/lbuchs/WebAuthn and storing the multiple keys which looks like this

```
-----BEGIN PUBLIC KEY-----
MFkwEwYHKoZIzj0CAQYIKoZIzj0DAQcDQgAE7Ll9i6DCs+XwO8MAVfT/5otferRB
NzTbNlJkKXktiYameLQK6dqIa8V/IoKo6noaQ65cQDbfPgCna4rDQSRECA==
-----END PUBLIC KEY-----

```

The internal workflow would be similar to this one: https://learnpasskeys.io/demo/create
And for the final users looks a bit like this: https://www.passkeys.io/

**Comment by aelaraji** [comment]
sounds great, but at least leave it as an option and let the the end users choose their preferred option. 

I myself relay on TOTP a lot, -even tho I still can't figure out how to set one up for Timeline- and to be honest, it may take me quite some time before adopting passkeys (if I ever do). 

But it wouldn't hurt having it as an option.

**Comment by eapl-gemugami** [comment]
About TOTP, the implementation is incomplete in `timeline`, it doesn't help you to create your secret.
In other PHP projects I have the following on-boarding: https://fit.eapl.me/sign_in.php (Use a fake email so you can see how the whole process).

And on Passkeys I think it's a matter of getting used to it. For instance I'm using a hardware token and also Vaultgarden so it's kind of easy to replace it everywhere. I don't use Google or Apple systems that much as I suppose most of the users will do.

**Comment by eapl-gemugami** [comment]
I drafted a proof of concept, you can try it here:
https://twtxt-sandbox.eapl.me/timeline/webauthn/register
https://twtxt-sandbox.eapl.me/timeline/webauthn/login

The workflow would be:
First login with the sandbox [password](https://github.com/sorenpeter/timeline/issues/18#issue-2721206444), then register your passkey in `webauthn/register`, then log out, and finally use `webauthn/login` to test it.

Please leave your comments :) !

**Reply by eapl-gemugami** [reply]
Did anyone have the chance to test it?

I'm watching that I need to fix some stuff like the empty state:
![image](https://github.com/user-attachments/assets/1deb545d-9236-4d94-b9c6-d8ac8cb81406)

And perhaps adding a longer timeout

**Reply by eapl-gemugami** [reply]
Also removing this warning when you the secrets.json doesn't exist yet
`Warning: file_get_contents(private/webauthn/secrets.json): Failed to open stream: No such file or directory in /homepages/4/d818294726/htdocs/sandbox.twtxt.eapl.me/timeline/libs/WebAuthn-2.2.0/JsonManager.php on line 31`

**Reply by aelaraji** [reply]
I've just given it a try using my Password manager. (KeepassXC)
- Logged-in with password
- Registered a new passkey
- logged out.

Upon visiting the [webauthn/login](https://twtxt-sandbox.eapl.me/timeline/webauthn/login) page, clicking the `Login with Passkey` button doesn't prompt me for a confirmation. I get a pop-up from my password manager instead, saying "**Error**: No logins found."

I thought it was something wrong with the password manager itself, but gave it a shot with github and it worked flawlessly.

Feel free to ping me if there's anything else I could help with,
Good luck!

**Reply by eapl-gemugami** [reply]
Thanks! 

Oops, I made a quick change to test Non resident keys and forgot to switch it back the version in the repo. It could break something.

I'll try with KeypassXC, since I've only tested with BitWarden and Google Password Manager.

Did you have to install some plug-in first?

**Reply by eapl-gemugami** [reply]
That seems to be the problem. I rolled back to the 'right' version, and also added the ID for KeePassXC
![image](https://github.com/user-attachments/assets/ac0316f5-8fbc-4a79-a5d4-5cba6eac8057)

I'll delete the keys in the server if you want to try again.

**Reply by aelaraji** [reply]
Nice! Just gave it another try and it worked fine this time! 👌

**Reply by aelaraji** [reply]
> Did you have to install some plug-in first?

For using Passkeys with KeepassXC?
if so, all you have to do to is going to the browser's KeepassXC extension's settings, scroll-down a bit there will be [Passkeys section with a check box for enabling it](https://keepassxc.org/docs/KeePassXC_UserGuide#_passkeys).

**Reply by eapl-gemugami** [reply]
Agreed! I couldn't make it work with Zen browser which didn't seem to support KeePassXC integration although with Firefox works well.

Thanks for your help!

**Reply by aelaraji** [reply]
> I couldn't make it work with Zen browser which didn't seem to support KeePassXC integration

How come? That's actually my main browser.
The Firefox integration option should work for Zen just fine. 

![2025-01-12_18-29-13_827x345](https://github.com/user-attachments/assets/f46a77a5-7aab-44e7-900c-bdc94e28366d)

Otherwise you could setting a custom browser configuration location: `Ctrl+, > Brwoser Integration > Advanced` there's an option all the way down for `Use a custom browser configuration location:` 
- Browser Type: Firefox
- Config Location: ~/.config/Mozilla/native-messaging-hosts/ (what I had by default)

![2025-01-12_18-30-22_1046x560](https://github.com/user-attachments/assets/2f7610f8-01e2-4dee-9cba-644963f4d18e)

**Reply by eapl-gemugami** [reply]
Interesting! Thanks, I'll take a look into it
